Add DisableExtensionOperations field

RadekManak opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**

/kind feature
<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:
This PR adds a new field to machine spec DisableExtensionOperations that when set to true ensures that no VMExtensions are configured on the machine by preventing users to add values to VMExtensions field in machine template and setting AllowExtensionOperations false in the instance osProfile.
Setting DisableExtensionOperations also disables the bootstrap failure detection VMExtension.

The intent of this change is to provide a way to create VMs that use images without the capability of running VMExtensions. Currently, these machines get stuck waiting for the bootstrapping VMExtension to get installed.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
<!-- Refer to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/developers/releasing.md#release-support for more information about which changes are eligible for backport -->

- [ ] cherry-pick candidate

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [X] squashed commits
- [ ] includes documentation
- [X] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add option DisableExtensionOperations to disable VMExtensions
```

/cc @damdo @JoelSpeed

JoelSpeed reviewed (commented):
I'm wondering if we should make this about disabling the bootstrapping extension rather than disabling all extensions, since we can already disable extensions by just adding an empty list?

Going even further, is this actually a three option case, no bootstrapping extension, the default bootstrapping extension, or a custom bootstrapping extension (if there's a difference between a bootstrapping extension and regular extension that is, though I'm not convinced it is?)

jackfrancis commented:
/assign @willie-yao

willie-yao commented:
/ok-to-test

willie-yao commented on api/v1beta1/azuremachine_types.go:
```diff
@@ -132,6 +132,11 @@ type AzureMachineSpec struct {
 	// +optional
 	DNSServers []string `json:"dnsServers,omitempty"`
 
+	// Specifies whether extension operations should be disabled on the virtual machine. This may only be set to True when no
+	// extensions are configured on the virtual machine.
+	// +optional
+	DisableExtensionOperations *bool `json:"disableExtensionOperations,omitempty"`
+
```
Yes the default extension is always added even if no extensions are specified. I'm wondering if there is a specific use-case for just disabling extension operations temporarily while `VMExtensions` is specified. If that's not something you need, adding the option to disable will probably be enough.

willie-yao commented on api/v1beta1/azuremachine_validation.go:
```diff
@@ -471,3 +476,14 @@ func ValidateCapacityReservationGroupID(capacityReservationGroupID *string, fldP
+	allErrs := field.ErrorList{}
+
+	if ptr.Deref(disableExtensionOperations, false) && len(vmExtensions) > 0 {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("AzureMachineTemplate", "spec", "template", "spec", "VMExtensions"), "VMExtensions must be empty when DisableExtensionOperations is true"))
```
Judging by the logic [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4792/files#diff-b6931ac61f37a13f66fc7836e039bf714a0553d5db4f313cc8ef31ca5f5647b9R378), it seems like this check is unnecessary. If the goal is to disable the default extension, a flag to explicitly do that would be better. That way, it's still possible to specify additional extensions without using the default one, and avoid confusion about the two conflicting fields.

RadekManak replied:
The intent of disableExtensionOperations is to disable all VMExtension for the machine. This  check is run in webhook to reject the machine template. Setting disableExtensionOperations true and configuring VMExtension would at best result with no VMExtension deplyoed (unexpteced behavior) and at worst machine failing to provision.

jackfrancis commented:
/test pull-cluster-api-provider-azure-e2e-optional

RadekManak commented:
My original idea with `DisableExtensionOperations` was to have an explicit toggle that the machine won't be running any VMExtensions that is visible in both the machine spec and Azure portal. I can see that what we actually care about is being able to disable the bootstrapping extension is just a side effect of that toggle can be confusing.

I have changed the PR to add `DisableBootstrappingVMExtension` instead. It controls only the creation of bootstrapping VMExtension. We can solve users being able to add VMExtensions that will fail by documenting that our images don't support them.

willie-yao commented:
> My original idea with DisableExtensionOperations was to have an explicit toggle that the machine won't be running any VMExtensions that is visible in both the machine spec and Azure portal.

Ah that makes sense. It might be good to log a warning in the webhook if user's do specify VMExtensions while also disabling the bootstapping extension since if they are disabling it, it probably means the VM does not support extensions.

willie-yao commented on api/v1beta1/azuremachine_types.go:
```diff
@@ -132,6 +132,12 @@ type AzureMachineSpec struct {
 
+	// DisableBootstrappingVMExtension disables CAPZ Bootstrapping VM extension.
+	// Its role is to detect and report Kubernetes bootstrap failure or success.
+	// Use this setting if VMExtensions are not supported by your image.
```
```suggestion
	// Use this setting only if VMExtensions are not supported by your image.
```
I think we want to change the wording here since it's not recommended to disable the extension. If there is another reason why someone would want to disable it then it's okay to leave it as is!

patrickdillon commented:
I was investigating this issue independently; so was happy to discover this PR. As is, the PR solves the problem we have with RHCOS being unable to run extensions. LGTM.

At risk of bikeshedding, if there is a chance that future extensions could be added by default, it might be better to relabel this `DisableDefaultVMExtensions` so that this represents disabling all default VM extensions rather than is specific to the bootstrap extension, which is the only default extension. That said, it seems unlikely to me that more extensions would be added.

patrickdillon commented:
> I don't think there will be more default VMExtensions in the future. I've kept it as is, unless maintainers disagree, then I can change it.

+1. This LGTM.

willie-yao commented:
+1 from me. I don't think we're planning on any future default extensions.

damdo commented:
@willie-yao who should we assign for approval?

jackfrancis commented on api/v1beta1/azuremachine_types.go:
```diff
@@ -132,6 +132,12 @@ type AzureMachineSpec struct {
+	// Its role is to detect and report Kubernetes bootstrap failure or success.
+	// Use this setting only if VMExtensions are not supported by your image.
+	// +optional
+	DisableBootstrappingVMExtension *bool `json:"disableBootstrappingVMExtension,omitempty"`
```
@RadekManak sorry for the confusion, I was following this effort from the edge but didn't make the effort to comment. I actually prefered the previous "disable all VM extensions" option. I think that is a more clarifying option to fulfill your problem, and would help other users in a similar scenario where VM extensions are not supported.

Disabling this particular extension doesn't make a lot of sense as you lose an explicit part of the CAPI bootstrapping contract guarantee; but it *does* make sense if your CAPZ VM scenario doesn't allow extensions that we want to ship a VM configuration with no Extensions in the configuration.

IMO

We'll prioritize merging this change w/ the previous implementation, thank you so much for this and for your patience!

RadekManak commented:
Changed it back to DisableExtensionOperations and added a note explaining that it should only be used when VMExtensions are not supported by the image and that it disables the bootstrapping extension.

willie-yao commented:
/retest

willie-yao commented on api/v1beta1/azuremachine_types.go:
```diff
@@ -132,6 +132,12 @@ type AzureMachineSpec struct {
+	// Use this setting only if VMExtensions are not supported by your image, as it disables CAPZ bootstrapping extension used for detecting Kubernetes bootstrap failure.
+	// This may only be set to True when no extensions are configured on the virtual machine.
+	// +optional
+	DisableExtensionOperations *bool `json:"disableExtensionOperations,omitempty"`
```
I'm thinking this field should be immutable since it doesn't make too much sense to disable extensions after the VM is already provisioned. Can you add an immutability check in `ValidateUpdate`? https://github.com/willie-yao/cluster-api-provider-azure/blob/5243ae7d5dae9483280e73da2904a3568fbf8839/api/v1beta1/azuremachine_webhook.go#L77

RadekManak replied:
Thanks for pointing this out. I thought that the entire spec was immutable and did not check. Fixed.

willie-yao reviewed (commented):
Looks good pending one comment!

willie-yao commented on api/v1beta1/azuremachine_webhook.go:
```diff
@@ -213,6 +213,13 @@ func (mw *azureMachineWebhook) ValidateUpdate(ctx context.Context, oldObj, newOb
 	}
 
+	if err := webhookutils.ValidateImmutable(
+		field.NewPath("spec", "disableExtensionOperations"),
```
```suggestion
		field.NewPath("spec", "DisableExtensionOperations"),
```

This should also be fixed for `capacityReservationGroupID` on line 210. Sorry for not catching this in the previous PR!

RadekManak replied:
Changed to use the capitalized variant. `capacityReservationGroupID` should be fixed in its own PR.

JoelSpeed replied:
Why with a capital? This breaks Kube API conventions. The path is supposed to take the json serialisation representation of the field name, so a `camelCase` is correct.

Adding `DisableExtensionOperations` with a capital means the error that comes back prints an incorrect JSON path

The `capacityReservationGroupID` is correct and should not be changed

willie-yao replied:
Hmm you're definitely correct. I'm seeing some inconsistencies in the use of `field.NewPath()` in CAPZ and that should definitely be fixed. Apologies for the confusion there!

RadekManak replied:
This PR is changed back to camelCase.

willie-yao commented:
/retest

willie-yao reviewed (approved):
/lgtm
/cc @jackfrancis for approval

jackfrancis commented:
/approve

Thank you @RadekManak @JoelSpeed!
